### PR TITLE
ENH: Adds record literals

### PR DESCRIPTION
**Pull request by llllllllll**
Implements the record literal syntax we discussed at strata

@cpcloud @jcrist

### CODE REVIEW

File: datashape/coretypes.py [hunk 1]
```diff
@@ -912,7 +914,37 @@ def __repr__(self):
...
+        return packed
+
+    def __getitem__(self, types):
+        if not isinstance(types, tuple):
+            types = types,
+
+        return self(map(self._unpack_slice, types, range(len(types))))
+
+
+class Record(RecordMeta('_', (CollectionPrinter, Mono), {})):
```
Comment (cpcloud) [inline datashape/coretypes.py]: any reason not to set `__metaclass__ = RecordMeta` in the class body of `Record`?

File: datashape/coretypes.py [hunk 2]
```diff
@@ -912,7 +914,37 @@ def __repr__(self):
...
+    @staticmethod
+    def _unpack_slice(s, idx):
+        if not isinstance(s, slice):
+            raise TypeError(
+                'invalid field specification at position %d.\n'
+                'fields must be formatted like: {name}::{type}' % idx,
+            )
+
+        name, type_ = packed = s.start, s.step
+        if name is None:
```
Comment (cpcloud) [inline datashape/coretypes.py]: can you add a check that this is an instance of `basestring`. you can import this from `compatibility` i believe

File: datashape/tests/test_coretypes.py [hunk 3]
```diff
@@ -539,3 +540,24 @@ def test_decimal_attributes():
...
     with pytest.raises(TypeError):
         d4 = Decimal()
+
+
+def test_record_literal():
+    assert R is Record
+    assert R['a'::'int32'] == R([('a', 'int32')])
+    assert R['a'::'int32', 'b'::'int64'] == R([('a', 'int32'), ('b', 'int64')])
+
+    invalids = (
```
Comment (cpcloud) [inline datashape/tests/test_coretypes.py]: can you make this a separate parametrized test?

File: datashape/coretypes.py (hunk 1 above)
Comment (llllllllll) [inline datashape/coretypes.py]: That doesn't work in python 3, in py3 you pass `metaclass=M` to the class like:

```python
class C(metaclass=M):
    pass
```

File: datashape/py2help.py [hunk 4]
```diff
@@ -45,6 +45,30 @@
...
+
+        # Pyton 2
+        class C(A, B):
+            __metaclass__ = M
+
+        # Python 3
+        class C(A, B, metaclass=M):
+            pass
     """
     return metaclass('_', bases, {})
```
Comment (cpcloud) [inline datashape/py2help.py]: Is there anything that relies on the name generated by the type constructor being unique?

Comment (llllllllll) [inline datashape/py2help.py]: I don't think so. It is valid to construct two classes with the same names in different modules (or even the same module) so this is fine.

**Comment by llllllllll** [comment]
Any other comments? Do you think that this is actually something we want in the library? I was working with some programatic dshapes today and it was pretty rough without this; however, I could understand not wanting to confuse people.

**Comment by cpcloud** [comment]
I'd mark this as an experimental feature then merge. Also add a big note that this particular syntax is subject change and possibly even removal

**Comment by llllllllll** [comment]
Where would I mark this, just in the whatsnew?

**Comment by llllllllll** [comment]
Also, one thought that I had when talking to scott was that this might be better with just a single `:` to look more like the dshape string.

**Comment by cpcloud** [comment]
I'm okay with either syntax, but not both.

**Comment by cpcloud** [comment]
Yep mark in what's new

**Comment by llllllllll** [comment]
Okay, I think I prefer the single then since it is closer to the dshape sytax. I will make that change and update the whatsnew.

**Comment by cpcloud** [comment]
Just merged a PR that probably conflicted with this one

**Comment by llllllllll** [comment]
kk, I will rebase. I don't think the cleanups will confilict in a difficult way.

**Pull request merged**